source/WeightedList.h

```diff
@@ -51,10 +51,10 @@ class WeightedList {
 	const_iterator end() const noexcept { return choices.end(); }
 
 	void clear() noexcept { choices.clear(); total = 0; }
-	std::size_t size() const noexcept { return choices.size(); }
+	int size() const noexcept { return choices.size(); }
 	bool empty() const noexcept { return choices.empty(); }
-	Type &back() noexcept { return choices.back(); }
-	const Type &back() const noexcept { return choices.back(); }
+	std::pair<Type, int> &back() noexcept { return choices.back(); }
+	const std::pair<Type, int> &back() const noexcept { return choices.back(); }
 
 	template <class ...Args>
 	Type &emplace_back(Args&&... args);
```

Comment:
I think it would make sense to require the first argument to be the weight, and then allow the remainder to be the arguments needed to construct `Type`.
e.g.
```c++
template <class ...Args>
Type &emplace_back(int weight, Args&&... args)
```

(You can then deduce the suitability of the weight before constructing Type, too)